cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp

```diff
@@ -356,30 +254,115 @@ sycl::event compute_derivative(sycl::queue& q,
         });
     });
+    return add_event;
+}
+
+template <typename Float>
+sycl::event add_regularization_gradient_loss(sycl::queue& q,
+                                             const ndview<Float, 1>& parameters,
+                                             ndview<Float, 1>& out,
+                                             ndview<Float, 1>& out_derivative,
+                                             Float L1,
+                                             Float L2,
+                                             bool fit_intercept,
+                                             const event_vector& deps) {
+    using dal::backend::operator+;
+    auto [reg_val, reg_val_e] = ndarray<Float, 1>::zeros(q, { 1 }, sycl::usm::alloc::device);
+
+    const std::int64_t p =
+        fit_intercept ? parameters.get_dimension(0) - 1 : parameters.get_dimension(0);
 
-    return reg_event;
+    const auto* const param_ptr = parameters.get_data();
+    auto* const reg_ptr = reg_val.get_mutable_data();
+    auto* const out_ptr = out.get_mutable_data();
+    auto* const grad_ptr = out_derivative.get_mutable_data();
+    auto new_deps = deps + reg_val_e;
+    auto reg_event = q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(new_deps);
+        const auto range = make_range_1d(p);
+        auto sum_reduction = sycl::reduction(reg_ptr, sycl::plus<>());
+        std::int64_t st_id = fit_intercept;
+        cgh.parallel_for(range, sum_reduction, [=](sycl::id<1> idx, auto& sum) {
+            const Float param = param_ptr[idx + st_id];
+            sum += L1 * sycl::abs(param) + L2 * param * param;
+            grad_ptr[idx + st_id] += L2 * 2 * param;
+        });
+    });
+
+    auto add_event = q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on({ reg_event });
+        cgh.single_task([=] {
+            out_ptr[0] += reg_ptr[0];
+        });
+    });
+
+    return add_event;
+}
+
+template <typename Float>
+sycl::event add_regularization_gradient(sycl::queue& q,
+                                        const ndview<Float, 1>& parameters,
+                                        ndview<Float, 1>& out_derivative,
+                                        Float L1,
+                                        Float L2,
+                                        bool fit_intercept,
+                                        const event_vector& deps) {
+    auto* const grad_ptr = out_derivative.get_mutable_data();
+    const auto* const param_ptr = parameters.get_data();
+    const std::int64_t p =
+        fit_intercept ? parameters.get_dimension(0) - 1 : parameters.get_dimension(0);
+    return q.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        const auto range = make_range_1d(p);
+        std::int64_t st_id = fit_intercept;
+        cgh.parallel_for(range, [=](sycl::id<1> idx) {
+            const Float param = param_ptr[idx + st_id];
+            grad_ptr[idx + st_id] += L2 * 2 * param;
+        });
```

Comment:
It can be done with an `element_wise` primitive as well

Reply:
As in different cases we update either whole array or just it is suffix of size p I find it more convenient to use manual kernel